torch/cuda/_memory_viz.py

```diff
@@ -824,6 +837,586 @@ def add_element(size, tensor_key, version):
+        }
+    }
+
+    def fold_free(ts):
+        # turn a free_requested/free_completed pair into a single free event
+        i = 0
+        while i < len(ts):
+            t = ts[i]
+            if i + 1 < len(ts):
+                tnext = ts[i + 1]
+                if t['action'] == 'free_requested' and tnext['action'] == 'free_completed' and t['addr'] == tnext['addr']:
+                    yield {**t, 'action': 'free'}
+                    i += 2
+                    continue
+            if t['action'] == 'oom':
+                yield {**t, 'addr': t['device_free']}
+            else:
+                yield t
+            i += 1
+
+    preproc: Any = {
+        'action': intern_str,
+        'frames': format_frames,
+    }
+
+    events: Any = result['events']
+    for event in fold_free(trace):
+        for k in events.keys():
+            # stack frames not recorded on event
+            # happens for snapshot even when
+            # frames are recorded for other things.
+            if k == 'frames' and k not in event:
+                events[k].append(None)
+                continue
+            events[k].append(preproc.get(k, lambda x: x)(event[k]))
+
+    segments = result['segments']
+    blocks = result['blocks']
+
+    segment_names = {
+        'addr': 'address',
+        'size': 'total_size',
+    }
+
+    for seg in data['segments']:
+        for k in segments.keys():
+            sk = segment_names.get(k, k)
+            segments[k].append(preproc.get(k, lambda x: x)(seg[sk]))
+        addr = seg['address']
+        for b in seg['blocks']:
+            if b['state'] in ('active_pending_free', 'active_allocated'):
+                if 'history' in b:
+                    frames = b['history'][0].get('frames', [])
+                    real_size = b['history'][0]['real_size']
+                else:
+                    real_size = b['size']
+                    frames = []
+                blocks['addr'].append(addr)
+                blocks['size'].append(b['size'])
+                blocks['real_size'].append(real_size)
+                blocks['frames'].append(format_frames(frames))
+                blocks['pending_free'].append(1 if b['state'] == 'active_pending_free' else 0)
+            addr += b['size']
+
+    plot_data = json.dumps(result)
+    return _events_template.replace('$PLOT_DATA', plot_data)
+
+_events_template = r"""
+<!DOCTYPE html>
+<html>
+<head>
+<style>
+pre {
+    margin: 0px;
+}
+html, body {
+    height: 100%;
+}
+</style>
+</head>
+<body>
+<script type="module">
+import * as d3 from "https://cdn.jsdelivr.net/npm/d3@7.7.0/+esm";
+import {schemeTableau10} from "https://cdn.skypack.dev/d3-scale-chromatic@3";
+import {axisLeft} from "https://cdn.skypack.dev/d3-axis@3";
+import {scaleLinear} from "https://cdn.skypack.dev/d3-scale@4";
+import {zoom, zoomIdentity} from "https://cdn.skypack.dev/d3-zoom@3";
+import {brushX} from "https://cdn.skypack.dev/d3-brush@3";
+
+let trace_data = $PLOT_DATA
```

Comment:
I'm guessing this (and others below) could probably be `const` rather than `let` (anything that's never reassigned, even if you mutate object props, array items etc.) 

ADDING JAVASCRIPT VALUE